### ISSUE TITLE
Mempool: Introduce event tracing

### ISSUE THREAD

**Issue by intricate** [input-output-hk/ouroboros-network#669]
As it stands, there is no tracing for `Mempool` events.

E.g. tracing type for `Mempool` update events from @dcoutts:
```
data TraceEventMempool blk =
       TraceMempoolAddTxs      [GenTx blk] Int
     | TraceMempoolRejectedTxs [GenTx blk] Int
     | TraceMempoolRemoveTxs   [GenTx blk] Int
```

**Comment by mrBliss** [comment]
@CodiePP you mentioned you want to have some mempool *metrics*, e.g., TPS. Can you say which metrics and trace events you'd like to see in the mempool?

**Comment by CodiePP** [comment]
We would like to see events traced when a new transaction is added to the mempool (reports the number of tx in mempool) and when a block is created how many tx remain in the mempool.

**Comment by mrBliss** [comment]
> ```
> data TraceEventMempool blk =
>        TraceMempoolAddTxs      [GenTx blk] Int Int
>      | TraceMempoolRejectedTxs [GenTx blk] Int Int
>      | TraceMempoolRemoveTxs   [GenTx blk] Int Int
> ```
We should use fields or newtypes for these `Int`s. Also, using `Word32/64` would rule out silly things like a negative number of transactions in the mempool.

**Comment by mrBliss** [comment]
> We would like to see events traced when a new transaction is added to the mempool (reports the number of tx in mempool) and when a block is created how many tx remain in the mempool.
> The above structure `TraceEventMempool` will probably report the number of tx that were added/removed/rejected from the mempool. This would mean that we had to do some bookkeeping. It would be easier to see events on the number of tx affected and the total number of tx in the mempool:
> 
> ```
> data TraceEventMempool blk =
>        TraceMempoolAddTxs      [GenTx blk] Int Int
>      | TraceMempoolRejectedTxs [GenTx blk] Int Int
>      | TraceMempoolRemoveTxs   [GenTx blk] Int Int
> ```

@CodiePP:

I'm assuming the difference between *rejected* and *removed* is the following:
* *rejected*: a new transaction that is submitted to the mempool is invalid and thus *rejected*.
* *removed*: a transaction was submitted to the mempool, was valid, and made it into the mempool. After synching with the updated ledger, that transaction is no longer valid and is thus *removed*.

Is it the convention to trace one event containing a list of things instead of tracing many events each containing a single thing? For example, trace one `TraceMempoolAddTxs [GenTx blk]` vs. many `TraceMempoolAddTx (GenTx blk)`. I don't know how these events are consumed, but in my mind, I would expect the latter to be "easier". On the producing side, it doesn't matter much.

**Comment by mrBliss** [comment]
Addressed by #696.
